File: JS/Bus.js
class Bus {
    /**
     * Bus can be initialized without any param 'let bus = new Bus()'
     * @param {object} options "default: {...}"
     * @param {string} options.Id "default: uuid"
     * @param {string} options.Address "default: window.location.host"
     * @param {string} options.Path "default: /ws/bus"
     * @param {boolean} options.Secure "default: false"
     * @param {boolean} options.Autorestart "default: false"
     * @param {number} options.RestartEvery "default: 10"
     */
    constructor(options) {
        if (options === undefined) {
            options = {}
        }
        this.Secure = options.Secure || false;
        this.Secure ? this.scheme = "wss://" : this.scheme = "ws://"
        this.Address = options.Address || window.location.host;
        this.Path = options.Path || "/ws/bus";
        this.fullAddress = this.scheme + this.Address + this.Path;
        this.TopicHandlers = {};
        this.Autorestart = options.Autorestart || false;
        this.RestartEvery = options.RestartEvery || 10;
        this.OnOpen = () => { };
        this.OnClose = () => { };
        this.OnDataWs = (data, ws) => { };
        this.OnId = (data) => { };
        this.Id = options.Id || this.makeid();
        this.conn = this.connect(this.fullAddress, this.callback);
    }

    connect(path, callbackOnData) {
        let $this = this;
        $this.conn = new WebSocket(path);
        $this.conn.binaryType = 'arraybuffer';
        $this.conn.onopen = (e) => {
            console.log("Bus Connected");
            $this.conn.send(JSON.stringify({
                "action": "ping",
                "from": $this.Id
            }));
            $this.TopicHandlers = {};
            $this.OnOpen();
        };

        $this.conn.onmessage = (e) => {
            let obj = JSON.parse(e.data);
            $this.subscription = {};
            $this.OnDataWs(obj, $this.conn);
            if (obj.event_id !== undefined) {
                $this.Publish(obj.event_id, {
                    "ok": "done",
                    "from": $this.Id
                })
            }
            if (obj.to_id !== undefined && obj.to_id === $this.Id && $this.OnId !== undefined) {
                delete obj.to_id
                $this.OnId(obj);
            }
            if (obj.topic !== undefined) {
                // on publish
                if ($this.TopicHandlers[obj.topic] !== undefined) {
                    let subs = new busSubscription($this, obj.topic);
                    $this.TopicHandlers[obj.topic](obj, subs);
                    return;
                } 
            }
        };

        $this.conn.onclose = (e) => {
            $this.OnClose();
            if ($this.Autorestart) {
                console.log('Socket is closed. Reconnect will be attempted in ' + this.RestartEvery + ' second.', e.reason);
                setTimeout(function () {
                    $this.conn = $this.connect(path, callbackOnData);
                }, this.RestartEvery * 1000);
            } else {
                console.log('Socket is closed:', e.reason);
            }
        };

        $this.conn.onerror = (err) => {
            console.log('Socket encountered error: ', err.message, 'Closing socket');
            $this.conn.close();
        };
        return $this.conn;
    }

    /**
     * Subscribe subscribe to a topic
     * @param {string} topic 
     * @param {function handler(data: string,subscription: busSubscription) {}} handler 
     */
    Subscribe(topic, handler) {
        this.conn.send(JSON.stringify({
            "action": "sub",
            "topic": topic,
            "from": this.Id
        }));
        let subs = new busSubscription(this, topic);
        this.TopicHandlers[topic] = handler;
        return subs;
    }

    /**
     * Unsubscribe unsubscribe from topic
     * @param {string} topic 
     */
    Unsubscribe(topic) {
        let data = {
            "action": "unsub",
            "topic": topic,
            "from": this.Id
        }
        this.conn.send(JSON.stringify(data));
        if (this.TopicHandlers !== undefined) {
            delete this.TopicHandlers[topic];
        }
    }

    /**
     * Publish publish to topic
     * @param {string} topic 
     * @param {object} data 
     */
    Publish(topic, data) {
        this.conn.send(JSON.stringify({
            "action": "pub",
            "topic": topic,
            "data": data,
            "from": this.Id
        }));
    }

    /**
     * PublishWaitRecv publish to topic and exec onRecv fn when recv data
     * @param {string} topic 
     * @param {object} data 
     * @param {function} onRecv 
     * @param {function} onExpire 
     */
    PublishWaitRecv(topic, data, onRecv, onExpire) {
        data.from = this.Id;
        data.topic = topic;
        let eventId = this.makeid();
        data.event_id = eventId;
        let done = false;

        let sub = this.Subscribe(eventId, (data, ch) => {
            done = true;
            if (onRecv) {
                onRecv(data);
            }
            ch.Unsubscribe();
        });
        this.Publish(topic, data);
        let timer = setTimeout(() => {
            clearTimeout(timer);
            if (!done) {
                if (onExpire) {
                    onExpire(eventId, topic);
                }
                sub.Unsubscribe();
            }
        }, 500);
    }

    /**
     * PublishToIDWaitRecv publish to topic and exec onRecv fn when recv data
     * @param {string} topic 
     * @param {object} data 
     * @param {function} onRecv 
     * @param {function} onExpire 
     */
    PublishToIDWaitRecv(id, data, onRecv, onExpire) {
        data.from = this.Id;
        data.id = id;
        let eventId = this.makeid();
        data.event_id = eventId;
        let done = false;

        let sub = this.Subscribe(eventId, (data, ch) => {
            done = true;
            if (onRecv) {
                onRecv(data);
            }
            ch.Unsubscribe();
        });
        this.PublishToID(id, data);
        let timer = setTimeout(() => {
            clearTimeout(timer);
            if (!done) {
                if (onExpire) {
                    onExpire(eventId, id);
                }
                sub.Unsubscribe();
            }
        }, 500);
    }



    /**
     * PublishToServer publish to a server using addr like localhost:4444 or domain name https
     * @param {string} addr 
     * @param {object} data 
     * @param {boolean} secure 
     */
    PublishToServer(addr, data, secure) {
        this.conn.send(JSON.stringify({
            "action": "pub_server",
            "addr": addr,
            "data": data,
            "secure": secure,
            "from": this.Id
        }));
    }

    /**
    * PublishToID publish to client or server id
    * @param {string} id 
    * @param {object} data 
    */
    PublishToID(id, data) {
        this.conn.send(JSON.stringify({
            "action": "pub_id",
            "id": id,
            "data": data,
            "from": this.Id
        }));
    }

    /**
     * RemoveTopic remove a topic completely from the server bus
     * @param {string} topic 
     * @returns 
     */
    RemoveTopic(topic) {
        if (topic !== "") {
            this.conn.send(JSON.stringify({
                "action": "remove_topic",
                "topic": topic,
                "from": this.Id
            }));
            return
        } else {
            console.error("RemoveTopic error: " + topic + " cannot be empty")
        }
    }

    makeid() {
        return "10000000-1000-4000-8000-100000000000".replace(/[018]/g, c =>
            (c ^ crypto.getRandomValues(new Uint8Array(1))[0] & 15 >> c / 4).toString(16)
        );
    }
}

/**
 * busSubscription is a class with one method allowing unsubscribing from a topic
 */
class busSubscription {
    constructor(cl, topic) {
        this.topic = topic;
        this.parent = cl;
    }
    /**
     * Unsubscribe take no params, unsubscribe from the topic
     */
    Unsubscribe() {
        this.parent.Unsubscribe(this.topic);
    }
}

File: PY/Bus.py
"""
This module implements a ksbus client https://github.com/kamalshkeir/ksbus

Author: Kamal Shkeir
Email: kamalshkeir@gmail.com
Link: https://kamalshkeir.dev
"""

import asyncio
import json
import random
import string

import websockets


class Bus:
    def __init__(self, options, block=False):
        self.addr = options.get('addr', 'localhost')
        self.path = options.get('path', '/ws/bus')
        self.scheme = 'ws://'
        if options.get('secure', False):
            self.scheme = 'wss://'
        self.full_address = self.scheme + self.addr + self.path
        self.conn = None
        self.topic_handlers = {}
        self.autorestart = options.get('autorestart', False)
        self.restartevery = options.get('restartevery', 5)
        self.OnOpen = options.get('OnOpen', lambda bus: None)
        self.OnClose = options.get('OnClose', lambda: None)
        self.OnDataWs = options.get('OnDataWs', None)
        self.OnId = options.get('OnId', lambda data: None)
        self.id = options.get('id') or self.makeId(12)
        try:
            if block :
                asyncio.get_event_loop().run_until_complete(self.connect(self.full_address))
            else:
                asyncio.create_task(self.connect(self.full_address))
        except Exception as e:
            print(e)
    
    async def connect(self, path):
        try:
            self.conn = await websockets.connect(path)
            await self.sendMessage({"action": "ping", "from": self.id})
            async for message in self.conn:
                obj = json.loads(message)
                if self.OnDataWs is not None:
                    self.OnDataWs(obj)
                if "event_id" in obj:
                    self.Publish(obj["event_id"], {"ok": "done", "from": self.id, "event_id":obj["event_id"]})
                if "to_id" in obj:
                    if self.OnId is not None:
                        self.OnId(obj)
                elif "topic" in obj:
                    if obj["topic"] in self.topic_handlers:
                        subs = BusSubscription(self, obj["topic"])
                        self.topic_handlers[obj["topic"]](obj, subs)
                    else:
                        print(f"topicHandler not found for: {obj['topic']}")
                elif "data" in obj and obj["data"] == "pong":
                    if self.OnOpen is not None:
                        self.OnOpen(self)
        except Exception as e:
            print(f"Server closed the connection: {e}")
            if self.autorestart:
                while True:
                    print(f"Reconnecting in {self.restartevery} seconds...")
                    await asyncio.sleep(self.restartevery)
                    await self.connect(self.full_address)

    def Subscribe(self, topic, handler):
        payload = {"action": "sub", "topic": topic, "from": self.id}
        subs = BusSubscription(self, topic)
        self.topic_handlers[topic] = handler
            
        if self.conn is not None:
            asyncio.create_task(self.sendMessage(payload))
        return subs

    def Unsubscribe(self, topic):
        payload = {"action": "unsub", "topic": topic, "from": self.id}
        del self.topic_handlers[topic]
            
        if topic and self.conn is not None:
            asyncio.create_task(self.sendMessage(payload))

    async def sendMessage(self, obj):
        try:
            await self.conn.send(json.dumps(obj))
        except Exception as e:
            print("error sending message:", e)

    def Publish(self, topic, data):
        if self.conn is not None:
            asyncio.create_task(self.sendMessage({"action": "pub", "topic": topic, "data": data, "from": self.id}))
        else:
            print("Publish: Not connected to server. Please check the connection.")

    def PublishToID(self, topic, data):
        if self.conn is not None:
            asyncio.create_task(self.sendMessage({"action": "pub_id", "id": self.id, "data": data, "from": self.id}))
        else:
            print("PublishToID: Not connected to server. Please check the connection.")

    def RemoveTopic(self, topic):
        if self.conn is not None:
            asyncio.create_task(self.sendMessage({"action": "remove", "topic": topic, "from": self.id}))
            del self.topic_handlers[topic]

    def makeId(self, length):
        return "".join(random.choices(string.ascii_letters + string.digits, k=length))

    def PublishWaitRecv(self, topic, data, onRecv, onExpire):
        data["from"] = self.id
        data["topic"] = topic
        eventId = self.makeId(8)
        data["event_id"] = eventId
        done = False

        self.Subscribe(eventId, lambda data, ch: self._onRecv(data, ch, onRecv, done))
        self.Publish(topic, data)
        asyncio.create_task(self._setExpireTimer(eventId, onExpire))

    async def _setExpireTimer(self, eventId, onExpire):
        await asyncio.sleep(0.5)
        if not self.topic_handlers.get(eventId):
            if onExpire:
                onExpire(eventId)

    def _onRecv(self, data,ch, onRecv, done):
        if not done:
            done = True
            if onRecv:
                onRecv(data)
                ch.Unsubscribe()

    def PublishToIDWaitRecv(self, id, data, onRecv, onExpire):
        data["from"] = self.id
        data["id"] = id
        eventId = self.makeId(8)
        data["event_id"] = eventId
        done = False

        self.Subscribe(eventId, lambda data, ch: self._onRecv(data, ch, onRecv, done))
        self.PublishToID(id, data)
        asyncio.create_task(self._setExpireTimer(eventId, onExpire))

    def PublishToServer(self, addr, data, secure):
        self.conn.send(json.dumps({
            "action": "pub_server",
            "addr": addr,
            "data": data,
            "secure": secure,
            "from": self.id
        }))


class BusSubscription:
    def __init__(self, bus, topic):
        self.bus = bus
        self.topic = topic

    async def sendMessage(self, obj):
        try:
            await self.bus.conn.send(json.dumps(obj))
        except Exception as e:
            print("error sending message:", e)

    def Unsubscribe(self):
        asyncio.create_task(self.sendMessage({"action": "unsub", "topic": self.topic, "from": self.bus.id}))
        del self.bus.topic_handlers[self.topic]

